pow/ethash.go

```diff
@@ -0,0 +1,572 @@
+
+			c.cache = make([]uint32, size/4)
+			generateCache(c.cache, c.epoch, seed)
+		}
+		// Iterate over all previous instances and delete old ones
+		for ep := int(c.epoch) - limit; ep >= 0; ep-- {
+			seed := seedHash(uint64(ep)*epochLength + 1)
+			path := filepath.Join(dir, fmt.Sprintf("cache-R%d-%x.%s", algorithmRevision, seed, endian))
+			os.Remove(path)
+		}
+	})
+}
+
+// release closes any file handlers and memory maps open.
+func (c *cache) release() {
+	if c.mmap != nil {
+		c.mmap.Unmap()
+		c.mmap = nil
+	}
+	if c.dump != nil {
+		c.dump.Close()
+		c.dump = nil
+	}
+}
+
+// dataset wraps an ethash dataset with some metadata to allow easier concurrent use.
+type dataset struct {
+	epoch uint64 // Epoch for which this cache is relevant
+
+	dump *os.File  // File descriptor of the memory mapped cache
+	mmap mmap.MMap // Memory map itself to unmap before releasing
+
+	dataset []uint32   // The actual cache data content
+	used    time.Time  // Timestamp of the last use for smarter eviction
+	once    sync.Once  // Ensures the cache is generated only once
+	lock    sync.Mutex // Ensures thread safety for updating the usage time
+}
+
+// generate ensures that the dataset content is generated before use.
+func (d *dataset) generate(dir string, limit int, test bool) {
+	d.once.Do(func() {
+		// If we have a testing dataset, generate and return
+		if test {
+			cache := make([]uint32, 1024/4)
+			generateCache(cache, d.epoch, seedHash(d.epoch*epochLength+1))
+
+			d.dataset = make([]uint32, 32*1024/4)
+			generateDataset(d.dataset, d.epoch, cache)
+
+			return
+		}
+		// If we don't store anything on disk, generate and return
+		csize := cacheSize(d.epoch*epochLength + 1)
+		dsize := datasetSize(d.epoch*epochLength + 1)
+		seed := seedHash(d.epoch*epochLength + 1)
+
+		if dir == "" {
+			cache := make([]uint32, csize/4)
+			generateCache(cache, d.epoch, seed)
+
+			d.dataset = make([]uint32, dsize/4)
+			generateDataset(d.dataset, d.epoch, cache)
+		}
+		// Disk storage is needed, this will get fancy
+		endian := "le"
+		if !isLittleEndian() {
+			endian = "be"
+		}
+		path := filepath.Join(dir, fmt.Sprintf("full-R%d-%x.%s", algorithmRevision, seed, endian))
+		logger := log.New("epoch", d.epoch)
+
+		// Try to load the file from disk and memory map it
+		var err error
+		d.dump, d.mmap, d.dataset, err = memoryMap(path, false)
+		if err == nil {
+			logger.Debug("Loaded old ethash dataset from disk")
+			return
+		}
+		logger.Debug("Failed to load old ethash dataset", "err", err)
+
+		// No previous dataset available, create a new cache file to fill
+		cache := make([]uint32, csize/4)
+		generateCache(cache, d.epoch, seed)
+
+		d.dump, d.mmap, d.dataset, err = memoryMapAndGenerate(path, dsize, func(buffer []uint32) { generateDataset(buffer, d.epoch, cache) })
+		if err != nil {
+			logger.Error("Failed generate mapped ethash dataset", "err", err)
+
+			d.dataset = make([]uint32, dsize/2)
```

Comment:
This would require extra code to track that generation failed and handle that in Search/Verify. Given that there's no meaningful way to get out of this problem, we might as well crash :P